File: app/src/main/java/com/castelijns/mmdemo/photos/PhotosRepo.java
package com.castelijns.mmdemo.photos;

import com.castelijns.mmdemo.app.SimpleCache;
import com.castelijns.mmdemo.models.Photo;
import com.castelijns.mmdemo.network.ApiManager;
import com.castelijns.mmdemo.network.ApiService;

import java.util.List;

import io.reactivex.Observable;

class PhotosRepo extends SimpleCache<Photo> {

    private static PhotosRepo instance = null;

    private static ApiService apiService;

    private PhotosRepo() {
        apiService = ApiManager.getService();
    }

    static PhotosRepo getInstance() {
        if (instance == null) {
            instance = new PhotosRepo();
        }

        return instance;
    }

    Observable<List<Photo>> getAllPhotos() {
        if (cache != null) {
            return Observable.just(cache);
        } else {
            return apiService.getAllPhotos();
        }
    }

    Observable<List<Photo>> getAllPhotosForAlbumId(int albumId) {
        return apiService.getAllPhotosForAlbumId(albumId);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/photos/PhotosRepo.kt
package com.castelijns.mmdemo.photos

import com.castelijns.mmdemo.app.SimpleCache
import com.castelijns.mmdemo.models.Photo
import com.castelijns.mmdemo.network.ApiManager
import com.castelijns.mmdemo.network.ApiService
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.coroutineScope

object PhotosRepo : SimpleCache<Photo>() {

    private var apiService: ApiService = ApiManager.service

    suspend fun getAllPhotos(): Deferred<List<Photo>> = coroutineScope {
        async { emptyList<Photo>() }
    }
//        get() = if (cache != null) {
//            Observable.just(cache)
//        } else {
//            apiService.getAllPhotos()
//        }

    suspend fun getAllPhotosForAlbumId(albumId: Int): Deferred<List<Photo>> =
            coroutineScope {
                async { emptyList<Photo>() }
//                async {
//                    apiService.getAllPhotosForAlbumId(albumId)
//                }
            }

}